libvast/src/system/terminator.cpp

```diff
@@ -54,49 +57,101 @@ caf::behavior terminator(caf::stateful_actor<terminator_state>* self) {
-    VAST_DEBUG(self, "got request to terminate", xs.size(), "actors");
-    VAST_ASSERT(!self->state.promise.pending());
-    self->state.promise = self->make_response_promise();
-    auto& remaining = self->state.remaining_actors;
-    remaining.reserve(xs.size());
-    for (auto i = xs.rbegin(); i != xs.rend(); ++i)
-      if (!*i)
-        VAST_DEBUG(self, "skips termination of already exited actor", i->id());
+  return {
+    [=](const std::vector<caf::actor>& xs) {
+      VAST_DEBUG(self, "got request to terminate", xs.size(), "actors");
+      VAST_ASSERT(!self->state.promise.pending());
+      self->state.promise = self->make_response_promise();
+      auto& remaining = self->state.remaining_actors;
+      remaining.reserve(xs.size());
+      for (auto i = xs.rbegin(); i != xs.rend(); ++i)
+        if (!*i)
+          VAST_DEBUG(self, "skips termination of already exited actor",
+                     i->id());
+        else
+          remaining.push_back(*i);
+      if (remaining.size() < xs.size())
+        VAST_DEBUG(self, "only needs to terminate", remaining.size(), "actors");
+      // Terminate early if there's nothing to do.
+      if (remaining.empty()) {
+        VAST_DEBUG(self, "quits prematurely because all actors have exited");
+        self->state.promise.deliver(atom::done_v);
+        self->quit(caf::exit_reason::user_shutdown);
+        return;
+      }
+      if constexpr (std::is_same_v<Policy, policy::sequential>) {
+        // Track actors in reverse order because the user provides the actors in
+        // the order of shutdown, but we use a stack internally that stores the
+        // first actor to be terminated at the end. In sequential operation, we
+        // monitor the next actor, send it an EXIT, wait for the DOWN, and then
+        // move to the next. This ensures that we'll always process the DOWN
+        // that corresponds to our EXIT message. (When monitoring an already
+        // terminated actor, CAF dispatches the DOWN immediately.)
+        auto& next = remaining.back();
+        self->monitor(next);
+        self->send_exit(next, caf::exit_reason::user_shutdown);
+      } else if constexpr (std::is_same_v<Policy, policy::parallel>) {
+        // Terminate all actors.
+        for (auto& x : xs) {
+          self->monitor(x);
+          self->send_exit(x, caf::exit_reason::user_shutdown);
+        }
+      } else {
+        static_assert(detail::always_false_v<Policy>, "unsupported policy");
+      }
+      // Send a reminder for killing all alive actors.
+      if (grace_period > std::chrono::milliseconds::zero())
+        self->delayed_send(self, grace_period, atom::shutdown_v);
+    },
+    [=](atom::shutdown) {
+      VAST_ASSERT(!self->state.remaining_actors.empty());
+      VAST_WARNING(self, "failed to terminate all actors within 10 mins");
+      VAST_WARNING(self, "initiates hard kill of",
+                   self->state.remaining_actors.size(), "remaining actors");
```

Comment:
These messages can be joined. Also, "10 mins" shouldn't be hardcoded here—the value is configurable now, isn't it?

Reply:
Full ACK regarding the hardcoding - this is fixed now.

While they can be joined, I'd like to keep log lines short. It doesn't hurt if you get a multi-line warning in my opinion.